### DISCUSSION TITLE
Tinder/StateMachine alternatives?

### DISCUSSION THREAD

**Discussion by mochadwi** [nsk90/kstatemachine#29] [Q&A] [answered]
Does `kstatemachine` inspired from `Tinder/StateMachine`? Would love to know about the comparisons between `kstatemachine` over `Tinder/StateMachine`

**Comment by nsk90** [comment, marked as answer]
Hi! No `kstatemachine` is not inspired by `Tinder/StateMachine` but DSL may look similar in some cases. `Kstatemachine` is intended to be similar by functionality to `QStateMachine` framework from `Qt` libraries, but  with nice Kotlin DSL syntax. I find `Qt` functionality more powerful than `Tinder/StateMachine`.

You can see `Qt` documentation, it is really good: https://doc.qt.io/qt-5/statemachine-api.html.

I am not a `Tinder/StateMachine` user, but I had a look at it's implementation when started this library. 
Here are some key differences that I know (maybe I am wrong in some cases):

* `Tinder/StateMachine` forces a user to process all transitions in a single function, `kstatemachine` not.
* `Tinder/StateMachine` does not support hierarchical state machines, `kstatemachine` does.
* `Tinder/StateMachine` does not support parallel states, `kstatemachine` does.
* `Tinder/StateMachine` does not support data states (that require some data in `Event` to become active), `kstatemachine` does.
* `Tinder/StateMachine` does not support conditional transitions (when target state is resolved during execution) , `kstatemachine` does.
* `State` class is immutable in `Tinder/StateMachine` but it is mutable in `kstatemachine`. This means that you cannot share State class instance across multiple StateMachines in `kstatemachine`.

See my wiki for more info: https://github.com/nsk90/kstatemachine/wiki

**Reply by mochadwi** [reply]
awesome! thanks for the answer, would consider to use kstatemachine in our codebase hopefully
